### DISCUSSION TITLE
2023.3.0 - Relative Translation, trackers and DCS; incorrect axis?

### DISCUSSION THREAD

**Discussion by Keebler750** [opentrack/opentrack#1805] [Q&A]
EDIT: I have modified the title of this thread to reflect the fact that ALL trackers seem to be affected, not just Tobii.

EDIT #2: It appears in part that DCS can duplicate tracker inputs and cause havoc. However, there is still an issue with what the expected behaviours are, and what documentation exists, especially for the Relative Translation page.

----------------------------------- original:

I'm LOVING the stability of tracking now that I'm using this setup, but there doesn't appear to be any documentation for the rather cryptic RT parameters and I'm having a tough time troubleshooting.

I have to ask - what is 'relative translation' supposed to be relative to? 

When using the native Tobii apps, relative translation is relative to the current view angle so that when you look to the side and move forward, your ingame view moves toward the view vector.

When using OpenTrack, the translations appear relative to the nose of the aircraft no matter what I try to do. When I look to the side and move my head forward as above, you can see the camera shifting toward the front of the plane, not toward the view vector.

Is this a setup problem on my end, or is this a known bug?

INPUT: Tobii Eye Tracker
OUTPUT: freetrack 2.0 enhanced; Track IR disabled.
FILTER: Hamilton
(Tobii Game Hub exited in System Tray)

**Comment by Keebler750** [comment]
After reading and researching, it is clear this is coded to flip some outputs depending on view angle. That is NOT the same as relative translation, and should not be labeled as such. Perhaps this has happened because the feature was requested based on 'looking behind an aircraft' instead of a more technical design related to camera position/rotation/zoom.

Relative translation is the binding of the head movements in the same way a human head works, namely: when you turn your head, and translate your head to your HEAD'S left, the view shifts relative to where it is pointed, not relative to the chair (or airplane) you are sitting in.

What this means is that the view vector becomes the reference, instead of the object we are piloting, and that is why the yaw/pitch/roll is processed first, and then translated in the RELATIVE x/y/z.

If you look at Tobii 5 output, their relative translation is coded as the human head would work, and it has been decoupled from the vehicle axis. This feels MUCH more natural and intuitive, since this is how we use our head and eyes in real life.

Something is happening in the OpenTrack code which continues to couple to the vehicle axis. The camera position output needs to couple to the view vector for actual 'relative translation.' 

I suspect this Discussion should be made an Issue.

**Comment by Keebler750** [comment]
I can report that even a Grass Monkey tracker works incorrectly in view translation. It's the underlying OpenTrack code, not the trackers. When you move your head toward the side of the aircraft you are looking out of, the view shifts to the FRONT of the aircraft, which is incorrect.

Why is this topic getting no response?

**Comment by CougarNL** [comment]
I'm using the (cap) point tracker and it works fine, maybe it's some trackers but not all of them.

Also gave the Tobii headtracker a go on the 2024 OT version and the relative translation works fine. Not that i would wanna use it, if i move my head too far to the side i end up with my face against the HUD. This could also be caused by the Tobii side of the head tracking or maybe the mapping in OT.

**Reply by CougarNL** [reply]
Also make sure you haven't disabled the Relative translation under options and try using the TrackIR interface only, no FreeTrack output.

**Reply by Keebler750** [reply]
So are you saying that when you turn your head toward the right wing tip, let's say, and then stick your chin (and the rest of your head) forward, the view goes toward the right wingtip, or does it shift toward the aircraft dash panel even though you're looking right? Because "Relative" translation should shift relative to where you're looking just like a real human head. Neither my Grass Monkey, nor my Tobii work as expected in this regard when using OpenTrack.... and I have experimented with every setting I could think of.

**Reply by CougarNL** [reply]
Yeah i know what relative translation is, couldn't fly without it :)

Vid: https://youtu.be/5fJMVKr6xGY

Are you on the 2024 OT build? Maybe something was fixed in that release. And maybe you can/should remove the Headtracker.dll from the DCS World OpenBeta\bin-mt folder, if it exists, to make sure you're not getting any FreeTrack input.

**Reply by Keebler750** [reply]
So.... I don't know what's going on. I REALLY want the Relative Translation to make sense. :s

headtracker.dll is NOT needed for Grass Monkey tracking with OpenTrack?

**Reply by CougarNL** [reply]
The headtracker.dll for DCS in that folder is for Freetrack input only. It's possible that other headtracker software uses it. With OT you can pick which output to use (Freetrack/TrackIR or both). The TrackIR output is usually the better choice for DCS.

Do you have 'freetrack 2.0 Enhanced' selected as output?

![OT_Output-001](https://github.com/user-attachments/assets/edb1da13-592e-4ef5-883d-60567b966fb4)

And for the settings under Select interface 'Use TrackIR, disable Freetrack'?

![FreeTrack-TrackIR-001](https://github.com/user-attachments/assets/6400f9fd-dcf7-4cac-9c4b-329fa9555267)

Options:

![OptionsRelTrans-001](https://github.com/user-attachments/assets/ce5ae4b7-4561-40e2-a17f-b74120ab6137)

And btw, in your OpenTrack, do you see the pink squid or a monkey?

**Reply by Keebler750** [reply]
Pink Squidly dude. 

I have OT set up exactly as above. RT does not work.... view camera moves toward front of aircraft.

**Reply by Keebler750** [reply]
Just discovered an issue: TWO trackers in the DCS "Controls" section. Leftover from Tobii? Tracker works totally different now, and I may be the victim/perp of another PEBCAK scenario. I REALLY appreciate you forcing me to look closer!

**Reply by CougarNL** [reply]
Yeah maybe it is some leftover from Tobii or something else. I would rename that Headtracker.dll to something else, Headtracker.dll.old or whatever, and restart DCS and check again.

**Reply by Keebler750** [reply]
Yeah, I renamed it, and then disabled the non-TrackIR tracker

**Reply by CougarNL** [reply]
And? Does it work now?

**Reply by Keebler750** [reply]
It's closer. I'm trying to tweak the setting and the curves. It must have been using the Tobii curves.

**Reply by CougarNL** [reply]
Cool :) For the rotation and position mapping i don't use any curves in OT. They're all just a diagonal lines from bottom left to top right (to keep it 1:1 and realistic).

**Reply by Keebler750** [reply]
Nope. Still not doing what I call relative translation. I'm going to talk to Grass Monkey before I continue, just to make sure I understand what this is SUPPOSED to do. The camera above my monitor, and OpenTrack, is not calculating the DIRECTION I'm moving my head. 

I mean... it's pretty good, but it's not right. The only relative translation I've seen so far is out of the stock Tobii software without OT.

**Reply by CougarNL** [reply]
Alright. And in OT? Does the squid move like it should? (turn head to the right, move head forward, squid moves to the side)

**Reply by Keebler750** [reply]
Nah. As I've said... we control the view vector which extends from the camera position. The translation ought to occur 'relative' to those axes, and it's not. What I'm seeing is the Tracker Camera is calculating relative to ITSELF, which is better than nothing, but the only 'Rotate view, then Translate' that I've seen is with the Tobii and NO OpenTrack .....which I thought was too jittery to use for DCS Apache targeting.

**Reply by Keebler750** [reply]
I'm building a 3D model for a video, and researching camera vector math.....ugh.....

**Reply by CougarNL** [reply]
Sounds like a lot of work when OT should work. Can you make a vid of what OT shows when you turn your head etc.etc?

**Reply by CougarNL** [reply]
Vid (fps is shitty because of desktop recording):
https://www.youtube.com/watch?v=yDb_bYT5qsw

And in DCS i only have 1 headtracking column, never touched those settings in DCS:

![DCS 2025-03-22 12-18-18-TrackIR](https://github.com/user-attachments/assets/e196943f-8b10-4af3-b0d4-3baab03b6fb6)

**Reply by Keebler750** [reply]
I went to my Hornet to be equal to yours, and right now it seems to be working properly. I deleted my Tobii drivers and stuff. Still have two columns for head tracking, but ... :shrug: it works. Hopefully this means I don't have to do any math LOL although I would like to understand what's going on.

**Reply by CougarNL** [reply]
As long as it works! :)

Not sure what math you mean, rotating stuff is like 3 lines of code but doing that in reverse from some dots on the screen is tricky. Had a peek at the FreeTrack source code years ago, that's just nuts.

Anywho, for the 2nd headtracker you could have a look in:
Saved Games\DCS.openbeta\Config\Input\FA-18C_hornet

I have 1 headtracker folder in that folder but it's empty (probably because i use default settings). Maybe you got 2 folders in that location? See if there's any files in those folders and if so maybe you can figure out what they're for. If they're there and they're lua files then you can open them with notepad.

**Reply by CougarNL** [reply]
Just noticed something, still tweaking and all. With the settings i posted, when i move my head forward and turn my head the rotation is around the origin. That aint right. These options seem to fix that.

![OptionsRelTransNew](https://github.com/user-attachments/assets/3239065b-88ed-4749-a92a-d62082f79567)

**Reply by CougarNL** [reply]
Never mind :P That kills the relative translation.

**Reply by CougarNL** [reply]
This would be that issue though. First 10 seconds i keep my head in the calibrated center, then i move my head forward and only rotate it... That's a litte off...

https://www.youtube.com/watch?v=frNefErNPOA

**Reply by Keebler750** [reply]
I think we need the camera in the middle of the monitor instead of on top! LOL. I notice that there are errors just like when you swing a mouse across a mousepad and induce Y axis changes when you are only looking for X. One reason I started playing from the elbow instead of the wrist. But I digress......

I'd love to see what factors are programmed into this tracker, because something is odd.

**Reply by CougarNL** [reply]
My cam is on a tripod behind the monitor looking down at an angle to make things even more fun!

Well, you could check out the source code. If there's no option to counter this then i think calculations aren't complete or the order in which stuff is calculated might be wrong.

**Reply by Keebler750** [reply]
I'm tempted to reinstall the Tobii and see if it too can work properly with OT. Without OT, the relative translation of that setup is BEAUTIFUL. The problem is that the Tobii doesn't have any proper filtering.

The Grass Monkey setup has made a huge improvement in the RT now with your help, but I'm interested....

**Reply by CougarNL** [reply]
Would be nice if OT would have 2 inputs, so you could select 1 for position and 1 for rotation.

For now i'll just stick to OT, in the pit i usually zoom instead of moving my head towards displays and gauges.

**Reply by Keebler750** [reply]
The Tobii has TWO IR sensors or something. I wonder if that gives a more 3 dimensional output?

SO much of the DCS ecosystem is based on stuff with little to no documentation. :(

**Reply by Keebler750** [reply]
When I had JUST the Tobii 5 installed as my first ever head tracker, IT didn't do relative translation in OpenTrack and DCS either. So...... what caused THAT instance of failure? There is not enough info to know!

**Reply by CougarNL** [reply]
The Tobii eyetracker 5 has 3 IR sensors and 1 camera. The IR sensors are for eye tracking and the cam is for head tracking.

DCS just accepts what you feed it (6 parameters, 6DOF) so users have the option to use relative translation or not.

**Reply by V1ctor-109** [reply]
> This would be that issue though. First 10 seconds i keep my head in the calibrated center, then i move my head forward and only rotate it... That's a litte off...
> 
> https://www.youtube.com/watch?v=frNefErNPOA

Sorry to disrupt you guys, but im curious if you're still around. I have the exact same issue as posted here by @CougarNL .
Although i can look at right wing and move forward towards the monitor to move my ingame view over the right wing, The issue is my view is rotating/orbiting along the center of the pilots seat no matter what i do. I cannot move forward or right and then rotate my view without ending up on the other side of the cockpit like headrest or the left window. 

I also made a post about this in #issues number #2170 but also to no avail from devs... 
But strangely i would swear OT had way better implementation of relative translation than this, but i can't for the life of me find on old version that would behave like i remember. It's really strange, do you have any more updates on your situations?
